### PR TITLE
feat: add an impl of `DoubleEndedIterator` for `Columns` and `Rows` (#1357)

### LINKED ISSUE #1357: Add an impl of `DoubleEndedIterator` for `Columns` and `Rows`
Add an impl of `DoubleEndedIterator` for `Columns` and `Rows`.

### PR DESCRIPTION
**Pull request by fujiapple852**
Closes #1357

### CODE REVIEW

File: src/layout/rect/iter.rs [hunk 1]
```diff
@@ -38,8 +38,10 @@ impl Iterator for Rows {
 pub struct Columns {
     /// The `Rect` associated with the columns.
     pub rect: Rect,
-    /// The x coordinate of the column within the `Rect`.
+    /// The x coordinate of the column within the `Rect` when iterating forwards.
     pub current_column: u16,
+    /// The x coordinate of the column within the `Rect` when iterating backwards.
+    current_column_back: u16,
```
Comment (fujiapple852) [inline src/layout/rect/iter.rs:44]: I chose not to make `current_column_back` pub, as I don' think it needs to be.  Given these are not exposed externally I would suggest renaming `current_column` as `current_column_fwd` and making it non-pub.

Comment (joshka) [reply src/layout/rect/iter.rs:44]: I don't recall why I made these pub in the first place - very likely by mistake rather than intention.
Changing this would be a breaking change, but one that very likely would be not seen by anyone.
I'd remove the pub altogether and slap a couple of lines in the docs.

File: src/layout/rect/iter.rs [hunk 2]
```diff
@@ -48,6 +50,7 @@ impl Columns {
         Self {
             rect,
             current_column: rect.x,
+            current_column_back: rect.x + rect.width,
```
Comment (joshka) [inline src/layout/rect/iter.rs:53]:
```suggestion
            current_column_back: rect.right(),
```

That said, I found it a little confusing that the current column is the next column to return, but the back column is the the next_back + 1 column. Does it make sense here to -1 on this and then use > instead of >= in the checks below?

e.g. for a 3 column area, this will currently be:
current_column = 0
current_column_back = 3 (should be 2)

File: src/layout/rect/iter.rs [hunk 3]
```diff
@@ -59,7 +62,9 @@ impl Iterator for Columns {
     ///
     /// Returns `None` when there are no more columns to iterate through.
     fn next(&mut self) -> Option<Self::Item> {
-        if self.current_column >= self.rect.right() {
+        if self.current_column >= self.rect.right()
+            || self.current_column >= self.current_column_back
```
Comment (joshka) [inline src/layout/rect/iter.rs:66]: given that current column back is initialized to right(), isn't the first check superfluos?
```suggestion
        if self.current_column >= self.current_column_back
```
(and change to `>` if the above minus 1 change is made

File: src/layout/rect/iter.rs [hunk 4]
```diff
@@ -68,6 +73,26 @@ impl Iterator for Columns {
     }
 }
 
+impl DoubleEndedIterator for Columns {
+    /// Retrieves the previous column within the `Rect`.
+    ///
+    /// Returns `None` when there are no more columns to iterate through.
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.current_column_back <= self.rect.x
+            || self.current_column_back <= self.current_column
```
Comment (joshka) [inline src/layout/rect/iter.rs:82]:
```suggestion
        if self.current_column_back <= self.current_column
```

(and change to `<` if needed)

**Review by joshka — APPROVED** [review]
LGTM generally. 
- We should break the use of the pub fields for 0.29.0
- Some small ergonomics of the code for symmetry / better understanding
- This PR should include rows() too

File: src/layout/rect/iter.rs (hunk 1 above)
Comment (fujiapple852) [reply src/layout/rect/iter.rs]: Done and added a section to the `BREAKING-CHANGES.md` document:

```
### Removed public field `[current_column|current_row]` from `layout::rect::[Columns|Rows]` ([#1358])

[#1358]: https://github.com/ratatui/ratatui/pull/1358

The `pub` modifier has been removed from fields `layout::rect::Columns::current_column` and 
`layout::rect::Rows::current_row`. These fields were not intended to be public and should not
have been accessed directly.
```

File: src/layout/rect/iter.rs (hunk 2 above)
Comment (fujiapple852) [reply src/layout/rect/iter.rs]: Changed to use `rect.right()`.

Things seem to work out badly when `current_column_back` points to the (and not past the) final index, mainly with the terminate case for `next_back()`.  I note that `right()` is defined as "the first coordinate outside of the `Rect`" which makes sense to me and seems consistent with this.  I tried to get it to work anyway but couldn't, maybe i'm just being stupid, feel free to have a fiddle.

File: src/layout/rect/iter.rs (hunk 3 above)
Comment (fujiapple852) [reply src/layout/rect/iter.rs]: Good catch, removed.

File: src/layout/rect/iter.rs (hunk 4 above)
Comment (fujiapple852) [reply src/layout/rect/iter.rs]: Likewise.

**Comment by fujiapple852** [comment]
Added `DoubleEndedIterator` for `Rows` as well, and added tests.

File: src/layout/rect/iter.rs (hunk 2 above)
Comment (joshka) [reply src/layout/rect/iter.rs]: Ah, you're right. The thing I was missing was that the indexes need to record when the iterators cross explicitly. A good simple case to test this is having a single item. If the indexes are both 0, then there's nothing to disambiguate whether the iterator has been exhausted already or if the iterator is in its initial state. Having them as 0, 1 makes this possible.

I wonder if there's problems around u16::MAX for this.

File: src/layout/rect/iter.rs [hunk 5]
```diff
@@ -114,20 +148,70 @@ mod tests {
...
+    #[test]
+    fn rows_back() {
+        let rect = Rect::new(0, 0, 2, 3);
+        let mut rows = Rows::new(rect).rev();
+        assert_eq!(rows.next(), Some(Rect::new(0, 2, 2, 1)));
+        assert_eq!(rows.next(), Some(Rect::new(0, 1, 2, 1)));
+        assert_eq!(rows.next(), Some(Rect::new(0, 0, 2, 1)));
+        assert_eq!(rows.next(), None);
+        assert_eq!(rows.next_back(), None);
+    }
```
Comment (joshka) [inline src/layout/rect/iter.rs:169]: I think this would be better to just test the iterator and not the RevIter wrapped here. Can this be:
```suggestion
    #[test]
    fn rows_back() {
        let rect = Rect::new(0, 0, 2, 3);
        let mut rows = Rows::new(rect);
        assert_eq!(rows.next_back(), Some(Rect::new(0, 2, 2, 1)));
        assert_eq!(rows.next_back(), Some(Rect::new(0, 1, 2, 1)));
        assert_eq!(rows.next_back(), Some(Rect::new(0, 0, 2, 1)));
        assert_eq!(rows.next_back(), None);
        assert_eq!(rows.next, None);
    }
```
And same for cols?

**Review by joshka — APPROVED** [review]
LGTM (repeating for other maintainers), hold merging this closer to 0.29 - I'd guess this might be sometime in ~2-4 weeks.

File: src/layout/rect/iter.rs (hunk 2 above)
Comment (fujiapple852) [reply src/layout/rect/iter.rs]: > I wonder if there's problems around u16::MAX for this.

I added a couple of boundary condition tests and I think it is correct.

```rust
/// We allow a total of `65536` columns in the range `(0..=65535)`.  In this test we iterate
/// forward and skip the first `65534` columns, and expect the next column to be `65535` and
/// the subsequent columns to be `None`.
#[test]
fn columns_max() {
    let rect = Rect::new(0, 0, u16::MAX, 1);
    let mut columns = Columns::new(rect).skip(usize::from(u16::MAX - 1));
    assert_eq!(columns.next(), Some(Rect::new(u16::MAX - 1, 0, 1, 1)));
    assert_eq!(columns.next(), None);
}

/// We allow a total of `65536` columns in the range `(0..=65535)`.  In this test we iterate
/// backward and skip the last `65534` columns, and expect the next column to be `0` and the
/// subsequent columns to be `None`.
#[test]
fn columns_min() {
    let rect = Rect::new(0, 0, u16::MAX, 1);
    let mut columns = Columns::new(rect).rev().skip(usize::from(u16::MAX - 1));
    assert_eq!(columns.next(), Some(Rect::new(0, 0, 1, 1)));
    assert_eq!(columns.next(), None);
}
```

Comment (joshka) [reply src/layout/rect/iter.rs]: thanks :)
Usually these things are not a huge problem, but someone somewhere is likely to try to break some random tui with this sort of thing.

**Review by orhun — APPROVED** [review]

**Review by joshka — APPROVED** [review]

File: src/layout/rect/iter.rs (hunk 1 above)
Comment (joshka) [reply src/layout/rect/iter.rs]: I removed the pub on the rect fields too.
